aten/src/ATen/native/cuda/Activation.cu

```diff
@@ -0,0 +1,60 @@
+#include "ATen/ATen.h"
+#include "ATen/NativeFunctions.h"
+#include "ATen/Dispatch.h"
+#include "ATen/cuda/CUDAApplyUtils.cuh"
+#include "ATen/cuda/CUDATensorMethods.cuh"
+#include "ATen/cuda/CUDATypeConversion.cuh"
+#include "THCUNN/THCHalfAutoNumerics.cuh"
+
+namespace at { namespace native {
+
+template <typename scalar_t>
+void hardshrink_cuda_kernel(Tensor& out_tensor, Tensor& lambd_tensor) {
+  at::cuda::CUDA_tensor_apply2<scalar_t, scalar_t>(
+      out_tensor,
+      lambd_tensor,
+      [] __device__ (scalar_t& out_tensor_val,
+         scalar_t& lambd_tensor_val,
+         bool early_exit) {
+           if (out_tensor_val >= -lambd_tensor_val && out_tensor_val <= lambd_tensor_val) {
+             out_tensor_val = ScalarConvert<double, scalar_t>::to(0.0);
+           }
+  });
+}
+
+template <typename scalar_t>
+void hardshrink_backward_cuda_kernel(Tensor& out_tensor, Tensor& lambd_tensor, const Tensor& self) {
+  at::cuda::CUDA_tensor_apply3<scalar_t, scalar_t, scalar_t>(
+      out_tensor,
+      lambd_tensor,
+      self,
+      [] __device__ (scalar_t& out_tensor_val,
+         scalar_t& lambd_tensor_val,
+         scalar_t& self_val) {
+           if (self_val >= -lambd_tensor_val && self_val <= lambd_tensor_val) {
+             out_tensor_val = ScalarConvert<double, scalar_t>::to(0.0);
+           }
+  });
+}
+
+Tensor hardshrink_cuda(const Tensor & self, Scalar lambd) {
+  auto lambd_tensor = at::zeros_like(self).fill_(lambd);
+  auto out_tensor = self.clone();
```

Comment:
This is quite wasteful in terms of performance - there will be 2 memory accesses per element to clone the tensor, and another 2 accesses to actually apply hardshrink to it, instead of just 2 accesses for optimal implementation.

Reply:
Thanks a lot! This is brilliant! I will use 2 memory accesses instead!

Reply:
Also, you don't have to materialize lambda tensor (that adds another couple of memory accesses), lambda can remain a scalar, similarly to how it is done e.g. here https://github.com/pytorch/pytorch/blob/f8e83dc257fbebc6988a174e18b87a4c00801fbc/aten/src/ATen/native/cuda/Distributions.cu#L46

Reply:
I must have missed something, but why the lambda looks like a tensor  in [this case](https://github.com/pytorch/pytorch/blob/f8e83dc257fbebc6988a174e18b87a4c00801fbc/aten/src/ATen/native/cuda/Distributions.cu#L37-L55) ?

Reply:
You are right, that was a bad example, a better one is here (though it's not a pretty lambda :-))
https://github.com/pytorch/pytorch/blob/master/aten/src/THCUNN/Threshold.cu#L6-L22
https://github.com/pytorch/pytorch/blob/master/aten/src/THCUNN/Threshold.cu#L6-L22

Reply:
found a way to pass in scalar in ATen style lol